Tapscripts support

bigspider opened the pull request:
This PR adds initial support for taproot scripts, with the following descriptor inside `tr`'s `TREE` expressions:
- `pk`
- `multi_a`
- `sortedmulti_a`

Incidentally, this also fixes:
- When signing taproot transactions, return 64-byte signatures if the sighash_type field in the psbt for that input is omitted.


TODO:

- [x] Document the API change (tapleaf concatenated to pubkey when yielding signatures), and the signing approach (what keys will the app sign with?)
- [x] optimize stack usage in new code
- [x] Decide a safe depth limit for taptrees in `wallet.c::parse_tree`.
- [x] Add more extensive tests

Closes: #86

sgliner-ledger commented on src/handler/sign_psbt.c:
```diff
@@ -1918,68 +1918,70 @@ sign_sighash_schnorr_and_yield(dispatcher_context_t *dc,
+    // using the TRY/CATCH block to ensure zeroing sensitive data.
 
-                uint32_t sign_path[MAX_BIP32_PATH_STEPS];
+    do {  // block executed once, only to allow safely breaking out on error
```
If you removed the try catch you can also remove the `do .. while()` I guess :)

bigspider replied:
there are still a couple of places where I use break out of the loop (and I can't use `return` because I have to make sure I call `explicit_bzero` below).

bigspider commented on src/handler/sign_psbt.c:
```diff
@@ -367,9 +367,64 @@ static int get_amount_scriptpubkey_from_psbt(
+                  psbt_key_type == PSBT_OUT_TAP_BIP32_DERIVATION;
+    if (is_tap) {
+        uint64_t hashes_len;
+        if (!buffer_read_varint(&response_buf, &hashes_len) || hashes_len > UINT16_MAX) {
```
Completely rewritten in a later commit.

bigspider commented on src/handler/sign_psbt.c:
```diff
@@ -2329,12 +2326,58 @@ sign_transaction(dispatcher_context_t *dc,
+                        // compute tapleaf hash
+                        // TODO: refactor this in a new function
+                        uint8_t script[80];  // TODO: will need a bigger buffer
+                        buffer_t script_buf = buffer_create(script, sizeof(script));
```
`buffer_create` returns a `buffer_t` struct, not a pointer; so it can't be null

bigspider commented on src/common/wallet.c:
```diff
@@ -1750,6 +1791,72 @@ static int parse_script(buffer_t *in_buf,
+            return -1;
+        }
+
+        // the next character must be a comma
```
Oops, fixed comment in 1c934f7fd6427cb938ebf8f477dba0fb244020d6.

bigspider commented on tests/test_register_wallet.py:
```diff
@@ -91,9 +91,12 @@ def test_register_wallet_reject_header(client: Client):
     for invalid_name in [
         "",  # empty name not allowed
-        "Very long walletz",  # 17 characters is too long
+        too_long_name,  # 65 characters is too long
```
There's already a separate test for it (`test_register_wallet_with_long_name`); I added an assert on the `name` length to make it more explicit.  1c934f7fd6427cb938ebf8f477dba0fb244020d6

bigspider commented on src/crypto.c:
```diff
@@ -541,12 +529,78 @@ static int crypto_tr_lift_x(const uint8_t x[static 32], uint8_t out[static 65])
-// TODO: should it recycle pubkey also for the output (like crypto_tr_tweak_seckey below)?
-int crypto_tr_tweak_pubkey(uint8_t pubkey[static 32], uint8_t *y_parity, uint8_t out[static 32]) {
+// Computes a tagged hash according to BIP-340.
+// If data_len > 0, then data2 must be non-NULL and the `data` and `data2` arrays are concatenated.
```
Right, fixed. 1c934f7fd6427cb938ebf8f477dba0fb244020d6

bigspider commented on src/handler/lib/policy.c:
```diff
@@ -757,18 +768,120 @@ static int process_multi_sortedmulti_node(policy_parser_state_t *state, const vo
+
+    // bitvector of used keys (only relevant for sorting keys in SORTEDMULTI)
+    uint8_t used[BITVECTOR_REAL_SIZE(MAX_PUBKEYS_PER_MULTISIG)];
+    memset(used, 0, sizeof(memset));
```
Let's pretend this never happened. 1c934f7fd6427cb938ebf8f477dba0fb244020d6

bigspider commented on src/handler/lib/policy.c:
```diff
@@ -757,18 +768,120 @@ static int process_multi_sortedmulti_node(policy_parser_state_t *state, const vo
+    memset(used, 0, sizeof(memset));
+
+    for (int i = 0; i < policy->n; i++) {
+        uint8_t compressed_pubkey[33];
```
Agree in principle, but if I add a `COMPRESSED_PUBKEY_LENGTH`, then I need a `XONLY_PUBKEY_LENGTH` and why not a `SHA256_HASH_LENGTH` and I think some lines are going to get very very long...

bigspider commented on src/handler/sign_psbt.c:
```diff
@@ -1742,8 +1737,12 @@ static bool __attribute__((noinline)) compute_sighash_segwitv1(dispatcher_contex
+    // ext_flag
+    uint8_t ext_flag = placeholder_info->is_tapscript ? 1 : 0;
+    // annex is not supported
+    uint8_t annex_present = 0;
```
Changed in 1c934f7fd6427cb938ebf8f477dba0fb244020d6.

bigspider commented:
Rebased; addressed most comments in 1c934f7fd6427cb938ebf8f477dba0fb244020d6 and added another commit 156251309b387223cd0ce4d02c82967e2a499946 to fix a minor warning from sonarcloud.
